src/rt/atomic.rs

```diff
@@ -45,245 +121,603 @@ struct Store {
+        f: impl FnOnce(u64) -> Result<u64, E>,
+    ) -> Result<u64, E> {
+        // Track the load is happening in order to ensure correct
+        // synchronization to the underlying cell.
+        self.track_load(threads);
+
+        // Apply coherence rules.
+        self.apply_load_coherence(threads, index);
+
+        self.stores[index].first_seen.touch(threads);
+
+        let prev = self.stores[index].value;
+
+        match f(prev) {
+            Ok(next) => {
+                // Track a store operation happened
+                self.track_store(threads);
+
+                // Perform load synchronization using the `success` ordering.
+                self.stores[index].sync.sync_load(threads, success);
 
-        if let Err(e) = f(index) {
-            self.history.stores[index].sync.sync_load(threads, failure);
-            return Err(e);
+                // Store the new value, initializing with the `sync` value from
+                // the load. This is our (hacky) way to establish a release
+                // sequence.
+                let sync = self.stores[index].sync;
+                self.store(threads, sync, next, success);
+
+                Ok(prev)
+            }
+            Err(e) => {
+                self.stores[index].sync.sync_load(threads, failure);
+                Err(e)
+            }
         }
+    }
 
-        self.history.stores[index].sync.sync_load(threads, success);
+    fn apply_load_coherence(&mut self, threads: &mut thread::Set, index: usize) {
+        for i in 0..self.stores.len() {
+            // Skip if the is current.
+            if index == i {
+                continue;
+            }
 
-        let mut new = Store {
-            // Clone the previous sync in order to form a release sequence.
-            sync: self.history.stores[index].sync.clone(),
-            first_seen: FirstSeen::new(threads),
-            seq_cst: is_seq_cst(success),
-        };
+            // READ-READ coherence
+            if self.stores[i].first_seen.is_seen_by_current(threads) {
+                let mo = self.stores[i].modification_order;
+                self.stores[index].modification_order.join(&mo);
+            }
+
+            // WRITE-READ coherence
+            if self.stores[i].happens_before < threads.active().causality {
+                let mo = self.stores[i].modification_order;
+                self.stores[index].modification_order.join(&mo);
+            }
+        }
+    }
+
+    /// Track an atomic load
+    fn track_load(&mut self, threads: &thread::Set) {
+        assert!(!self.is_mutating, "atomic cell is in `with_mut` call");
 
-        new.sync.sync_store(threads, success);
-        self.history.stores.push(new);
+        let current = &threads.active().causality;
 
-        Ok(index)
+        assert!(
+            self.unsync_mut_at <= *current,
+            "Causality violation: \
+             Concurrent load and mut accesses"
+        );
```

Comment:
probably a follow-up but it would be good to print the call locations of the load and mutable access here?

Reply:
yes. The entire code base needs to have location tracking added. It can be done incrementally in follow up PRs.